### PR TITLE
feat: aws-machine-image datasource

### LINKED ISSUE #11863: Support for AWS Machine Images
### What would you like Renovate to be able to do?

I think it would be good, if renovate could create automatic merge requests for amazon machine image updates. 

We create every week new machine images that include the latest (kernel) updates and instead of patching existing machines, we switch to these new amis. 
Instead of relying on dynamic resolution of ssm parameters, or dynamic filtering the latest machine image while deploying an application, it would be really good, if renovate could create a merge request for me with the new ami id.

So what i want to achieve is getting mergerequests for ami ids as i get for docker image tags (even if i know, this has to be handled in another way).

I think i am able to implement this, but would like to get feedback, if you also think this is a useful feature and my thoughts of how to implementing this find your support.

### If you have any ideas on how this should be implemented, please tell us here.

I think, we have to implement a new datasource for resolving aws images by a given filter ... 

and than we have to implement a new manager (or can use the regex manager?) ... I think we should allow all filters, that are supported via the aws api:

```
architecture - The image architecture (i386 | x86_64 | arm64).
block-device-mapping.delete-on-termination - A Boolean value that indicates whether the Amazon EBS volume is deleted on instance termination.
block-device-mapping.device-name - The device name specified in the block device mapping (for example, /dev/sdh or xvdh).
block-device-mapping.snapshot-id - The ID of the snapshot used for the Amazon EBS volume.
block-device-mapping.volume-size - The volume size of the Amazon EBS volume, in GiB.
block-device-mapping.volume-type - The volume type of the Amazon EBS volume (io1 | io2 | gp2 | gp3 | sc1 | st1 | standard).
block-device-mapping.encrypted - A Boolean that indicates whether the Amazon EBS volume is encrypted.
description - The description of the image (provided during image creation).
ena-support - A Boolean that indicates whether enhanced networking with ENA is enabled.
hypervisor - The hypervisor type (ovm | xen).
image-id - The ID of the image.
image-type - The image type (machine | kernel | ramdisk).
is-public - A Boolean that indicates whether the image is public.
kernel-id - The kernel ID.
manifest-location - The location of the image manifest.
name - The name of the AMI (provided during image creation).
owner-alias - The owner alias (amazon | aws-marketplace). The valid aliases are defined in an Amazon-maintained list. This is not the Amazon Web Services account alias that can be set using the IAM console. We recommend that you use the Owner request parameter instead of this filter.
owner-id - The Amazon Web Services account ID of the owner. We recommend that you use the Owner request parameter instead of this filter.
platform - The platform. To only list Windows-based AMIs, use windows.
product-code - The product code.
product-code.type - The type of the product code (marketplace).
ramdisk-id - The RAM disk ID.
root-device-name - The device name of the root device volume (for example, /dev/sda1).
root-device-type - The type of the root device volume (ebs | instance-store).
state - The state of the image (available | pending | failed).
state-reason-code - The reason code for the state change.
state-reason-message - The message for the state change.
sriov-net-support - A value of simple indicates that enhanced networking with the Intel 82599 VF interface is enabled.
tag:<key> - The key/value combination of a tag assigned to the resource. Use the tag key in the filter name and the tag value as the filter value. For example, to find all resources that have a tag with the key Owner and the value TeamA, specify tag:Owner for the filter name and TeamA for the filter value.
tag-key - The key of a tag assigned to the resource. Use this filter to find all resources assigned a tag with a specific key, regardless of the tag value.
virtualization-type - The virtualization type (paravirtual | hvm).
```
### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by markussiebert**
## Changes:

Introducing the aws-machine-image datasource

## Context:

Closes #11863

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
@markus7811 could you fix your local git user so that it corresponds with your GitHub account and then force push to your branch over the existing commit? That way the CLA bot can determine that you're not committing someone else's work :)

**Comment by markus7811** [comment]
can you change this PR to Draft? It's far from getting merged - only wanted to "visualize" my thoughts and get a deeper understanding of how this could be implemented ...

**Comment by rarkins** [comment]
I think you have access to converting it too. Look for the "Convert to draft" link (it's not a button)

**Comment by HonkingGoose** [comment]
> I think you have access to converting it too. Look for the "Convert to draft" link (it's not a button)

You're correct, quote from the GitHub docs:

> People with write permissions to a repository and pull request authors can change the stage of a pull request.

Read more here: https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request

**Comment by markus7811** [comment]
Sorry - till today I was not able to makethis WIP - but maybe you did?
image
Sorry - I am running a larger Gitlab Instance, but github.com is really new to me :-/

**Comment by rarkins** [comment]
Yes, I converted it to draft when you asked. Just wanted to let you know that you should also have the ability to do that directly in future whenever you want

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:1]:
```suggestion
This datasource returns the latest [Amazon Machine Image](https://docs.aws.amazon.com/en_en/AWSEC2/latest/UserGuide/AMIs.html) via the AWS API (valid credentials required).
```

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:3]:
```suggestion
Because there is no general lookupName, you have to use the [describe images filter](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-ec2/interfaces/describeimagescommandinput.html#filters) as minified JSON as a lookupName.
```

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:8]:
```suggestion
# Getting the latest official EKS image from AWS (account '602401143452' for eu-central-1) for EKS 1.21 (name matches 'amazon-eks-node-1.21-*') would look as a describe images filter like:
```

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md]:
```suggestion
At the moment, this datasource has no "manager".
You should use the regex manager for this - an example how a regex manager could look like:
```

Use one sentence per line. 😉

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:30]:
```suggestion
At the moment, this datasource has no "manager".
You have to use the regex manager for this.
```

Use one sentence per line.

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:51]:
```suggestion
This would match every file, and would recognize the following lines:
```

File: lib/versioning/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/versioning/aws-machine-image/readme.md:1]:
```suggestion
Renovate's AWS Machine Image versioning is a kind of hack to support Amazon Machine Images (AMI) updates.
```

File: lib/versioning/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/versioning/aws-machine-image/readme.md:3]:
```suggestion
At the moment every AMI that matches the regex `^ami-[a-z0-9]{17}$` is considered a valid "release".
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/datasource/aws-machine-image/readme.md
Comment (markus7811) [reply lib/datasource/aws-machine-image/readme.md]: Yes, thank you!

**Comment by markus7811** [comment]
I run ```yarn doc-fix``` but it changed nothing. Apart from that, I fixed a lot in the tests (coverage errors) and so on. So maybe you can run the checks again?

**Comment by HonkingGoose** [comment]
> I run `yarn doc-fix` but it changed nothing.

Oh okay, it was alright then. 😄 

> Apart from that, I fixed a lot in the tests (coverage errors) and so on. So maybe you can run the checks again?

Try running `yarn jest`, to see if your changes passes the tests now.

File: lib/datasource/aws-machine-image/index.ts
Comment (viceice) [inline lib/datasource/aws-machine-image/index.ts:1]: Why?

File: lib/datasource/aws-machine-image/index.ts
Comment (viceice) [inline lib/datasource/aws-machine-image/index.ts:6]:
```suggestion
import type { GetReleasesConfig, ReleaseResult } from '../types';
```

File: lib/versioning/aws-machine-image/index.ts
Comment (viceice) [inline lib/versioning/aws-machine-image/index.ts:6]: Remove

File: package.json
Comment (viceice) [inline package.json:204]: Why is this needed?

File: lib/versioning/aws-machine-image/index.ts
Comment (markus7811) [reply lib/versioning/aws-machine-image/index.ts:6]: of course, thank you!

File: lib/datasource/aws-machine-image/index.ts
Comment (markus7811) [reply lib/datasource/aws-machine-image/index.ts:6]: yes!

File: lib/datasource/aws-machine-image/index.ts
Comment (markus7811) [reply lib/datasource/aws-machine-image/index.ts:1]: sorry - artifact of my early development. Of course this can be removed! Thank you!

File: package.json
Comment (markus7811) [reply package.json:204]: It's a peer dependency of the aws api mock packe I used for my tests... and i don't now which it was lint or test complained that the peer dependency was not found... i try to reproduce it.

Comment (markus7811) [reply package.json:204]:
```
$ tsc --noEmit
node_modules/aws-sdk-client-mock/dist/types/libStorage.d.ts:8:95 - error TS2307: Cannot find module '@aws-sdk/client-s3' or its corresponding type declarations.

8 export declare const mockLibStorageUpload: (s3Mock?: import("./awsClientStub").AwsStub<import("@aws-sdk/client-s3").ServiceInputTypes, import("@aws-sdk/client-s3").ServiceOutputTypes> | undefined) => import("./awsClientStub").AwsStub<import("@aws-sdk/client-s3").ServiceInputTypes, import("@aws-sdk/client-s3").ServiceOutputTypes>;
                                                                                                ~~~~~~~~~~~~~~~~~~~~

node_modules/aws-sdk-client-mock/dist/types/libStorage.d.ts:8:143 - error TS2307: Cannot find module '@aws-sdk/client-s3' or its corresponding type declarations.

8 export declare const mockLibStorageUpload: (s3Mock?: import("./awsClientStub").AwsStub<import("@aws-sdk/client-s3").ServiceInputTypes, import("@aws-sdk/client-s3").ServiceOutputTypes> | undefined) => import("./awsClientStub").AwsStub<import("@aws-sdk/client-s3").ServiceInputTypes, import("@aws-sdk/client-s3").ServiceOutputTypes>;
                                                                                                                                                ~~~~~~~~~~~~~~~~~~~~

node_modules/aws-sdk-client-mock/dist/types/libStorage.d.ts:8:242 - error TS2307: Cannot find module '@aws-sdk/client-s3' or its corresponding type declarations.

8 export declare const mockLibStorageUpload: (s3Mock?: import("./awsClientStub").AwsStub<import("@aws-sdk/client-s3").ServiceInputTypes, import("@aws-sdk/client-s3").ServiceOutputTypes> | undefined) => import("./awsClientStub").AwsStub<import("@aws-sdk/client-s3").ServiceInputTypes, import("@aws-sdk/client-s3").ServiceOutputTypes>;
                                                                                                                                                                                                                                                   ~~~~~~~~~~~~~~~~~~~~

node_modules/aws-sdk-client-mock/dist/types/libStorage.d.ts:8:290 - error TS2307: Cannot find module '@aws-sdk/client-s3' or its corresponding type declarations.

8 export declare const mockLibStorageUpload: (s3Mock?: import("./awsClientStub").AwsStub<import("@aws-sdk/client-s3").ServiceInputTypes, import("@aws-sdk/client-s3").ServiceOutputTypes> | undefined) => import("./awsClientStub").AwsStub<import("@aws-sdk/client-s3").ServiceInputTypes, import("@aws-sdk/client-s3").ServiceOutputTypes>;
```

Thats the error  I get without this dev dependency...

**Comment by rarkins** [comment]
I'm seeing a conflict in package.json for now

**Comment by markus7811** [comment]
fixed liniting issue and merge conflict

**Review by rarkins — CHANGES_REQUESTED** [review]
Great docs! Is there an example repo you can share showing it in action?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by HonkingGoose — DISMISSED** [review]
Docs look okay to me now. 😄

**Comment by rarkins** [comment]
Needs coverage fix:

image

**Comment by markussiebert** [comment]
So I think I fixed all code coverage issues. ;-) Didn't managed to get those steps running in my forked repo.

File: package.json
Comment (markussiebert) [reply package.json]: So we can remove it? Or not? @rarkins you added it again, after it was removed?

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:9]: Please split this out to a separate pr.

File: package.json
Comment (viceice) [inline package.json:206]: remove it if it's only a unused peer dependency, the install warning can be ignored

**Comment by viceice** [comment]
Missing coverage

File: package.json
Comment (markussiebert) [reply package.json:208]: ok, removed

File: lib/util/template/index.ts
Comment (markussiebert) [reply lib/util/template/index.ts:9]: I would like to keep this in here, because I would like to use it in my "personal" prs to include a pretty representation of the lookup name in my pull requests. As you can see, the lookup name is a string representation of JSON... so I think this is related to this PR.
I implemented this earlier, but than removed it because @rarkins wanted to keep the default behavior for the DataSource. But I would like to use it in my personal settings - but think there it would not be possible extend this functionality via config.

**Comment by markussiebert** [comment]
@viceice how can I see where are coverage issues? I have no idea how to check this on my local machine

File: package.json
Comment (markussiebert) [reply package.json]: image
=> So I get this error after removing the dependency... now typecheck is failing :-/ sorry - I am a little bit lost with this

Comment (markussiebert) [reply package.json]: that was the reason I added the dependency

**Comment by rarkins** [comment]
> @viceice how can I see where are coverage issues? I have no idea how to check this on my local machine

If you run `yarn jest` locally then it creates a `coverage/` directory with a html index file you can load locally to see the gaps

File: lib/util/template/index.ts
Comment (viceice) [reply lib/util/template/index.ts:9]: I like to split this out to a separate PR to be merged before this.

So simply open a new PR with this change only. Then we merge that and afterwards you can merge main again and this will be resolved. 🤗

**Review by viceice — COMMENTED** [review]
Ok, readd the S3 client as dev dep, if it's required for typescript compile. 🤗

File: lib/util/template/index.ts
Comment (markussiebert) [reply lib/util/template/index.ts:9]: Ok, here we go: https://github.com/renovatebot/renovate/pull/12643

**Comment by markussiebert** [comment]
So next try :-) thank you @rarkins for your hint concerning the test coverage. Now it was easier for me to increase the test coverage and I reworked the whole tests (and parts of the datasource).

The tests that are failing now are not part of my changes so I am unsure, if I should fix them or not?

File: lib/datasource/artifactory/__snapshots__/index.spec.ts.snap
Comment (viceice) [inline lib/datasource/artifactory/__snapshots__/index.spec.ts.snap:8]: Revert this file

Comment (markussiebert) [reply lib/datasource/artifactory/__snapshots__/index.spec.ts.snap]: oh, of course - that change was done unintentionally

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:5]:
```suggestion
Because there is no general `lookupName`, you have to use the [describe images filter](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-ec2/interfaces/describeimagescommandinput.html#filters) as minified JSON as a `lookupName`.
```

Do we want to use the monospaced font for `lookupName` maybe?

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:25]: The YAML takes up a bit of horizontal space, can this be condensed in some way maybe?

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:54]:
```suggestion
Or as JSON:
```

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:32]:
```suggestion
At the moment, AMI files do not have a "manager".
```

Maybe this is more accurate?

File: lib/datasource/aws-machine-image/readme.md
Comment (HonkingGoose) [inline lib/datasource/aws-machine-image/readme.md:37]:
```suggestion
Here's an example of using the regex manager:
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
The docs need a review from somebody who actually knows about Amazon Machine Images, I cannot say if the docs are technically accurate or not...

File: lib/datasource/aws-machine-image/readme.md
Comment (markussiebert) [reply lib/datasource/aws-machine-image/readme.md]: I changed it to a more compact style ...

**Comment by markussiebert** [comment]
@rarkins sorry, but I see that there are changes you requested, but can't find them...

**Review by rarkins — DISMISSED** [review]

**Comment by rarkins** [comment]
@viceice please merge if you're happy with this

File: lib/datasource/aws-machine-image/index.ts
Comment (viceice) [inline lib/datasource/aws-machine-image/index.ts:31]:
```suggestion
  private readonly ec2: EC2Client;
```

File: lib/datasource/aws-machine-image/index.ts
Comment (viceice) [inline lib/datasource/aws-machine-image/index.ts:46]: Do we really need to have this public? can't we simply test it via the normal public api?

File: lib/datasource/aws-machine-image/index.spec.ts
Comment (viceice) [inline lib/datasource/aws-machine-image/index.spec.ts:7]: @HonkingGoose  Can you open a new issue to refactor aws tests, so all tests which use the aws-sdk are using the `mockClient` or not

https://github.com/renovatebot/renovate/blob/b4a84065a1e658a7be637c7b381d8828dea80a3a/lib/datasource/docker/index.spec.ts#L18

**Review by viceice — COMMENTED** [review]
some small suggestions

File: lib/datasource/aws-machine-image/index.ts
Comment (viceice) [reply lib/datasource/aws-machine-image/index.ts:46]: ☝️ ❓

**Review by viceice — COMMENTED** [review]
My Question isn't answered 🤷‍♂️

File: lib/datasource/aws-machine-image/index.spec.ts
Comment (HonkingGoose) [reply lib/datasource/aws-machine-image/index.spec.ts:7]: Made a new issue, see: #12739

File: lib/datasource/aws-machine-image/index.ts
Comment (HonkingGoose) [reply lib/datasource/aws-machine-image/index.ts:46]: Hi @markussiebert Can you respond to the questions raised by the maintainer/reviewer? I'll add the link so you can find it easily, it's easy to miss in the sea of review comments:

https://github.com/renovatebot/renovate/pull/11869#discussion_r750115854

Comment (markussiebert) [reply lib/datasource/aws-machine-image/index.ts:46]: omg, sorry ... I answered, but had to submit my answers as a review? This differs so much from what I am used to with gitlab ...

Comment (viceice) [reply lib/datasource/aws-machine-image/index.ts:46]: I think the cache decorator should also work on private methods as there is no difference to public at runtime.

So this is only required by direct testing, which is ok for now, as i makes tests easier.

BTW: Caching is disabled on tests and don't need to be tested. We have some special tests to verify caching 😉

**Review by viceice — DISMISSED** [review]

File: lib/datasource/aws-machine-image/index.ts
Comment (markussiebert) [reply lib/datasource/aws-machine-image/index.ts:46]: @viceice  I tried out locally to make this method private. Codecoverage was still 100%, but the decorator wasn't working anymore.

Comment (viceice) [reply lib/datasource/aws-machine-image/index.ts:46]: @JamieMagee do you why it's not working for private methods?

Comment (JamieMagee) [reply lib/datasource/aws-machine-image/index.ts:46]: You're correct that private modifiers don't exist at runtime, but the typescript compiler still enforces them.

For example, I've got the `jenkins-plugin` datasource open right now. If I make one of the decorated methods private, I get this error: `Argument of type '"getJenkinsPluginInfo"' is not assignable to parameter of type 'keyof JenkinsPluginsDatasource'.ts(2345)`

That's because of this code:

https://github.com/renovatebot/renovate/blob/789caadbb49e232ef74f59f5c41c1a6fbffee5c0/lib/util/cache/package/decorator.ts#L6-L10

The decorator can decorate methods that are in `keyof U`, in this case `keyof JenkinsPluginsDatasource`.

I can't think of any way that would allow us to decorate private methods, but still retain type safety.

Comment (JamieMagee) [reply lib/datasource/aws-machine-image/index.ts:46]: There is an open discussion in the TypeScript repo about this scenario https://github.com/microsoft/TypeScript/issues/22677

**Review by rarkins — APPROVED** [review]

**Pull request merged**
